why-cant-we.md

```diff
@@ -0,0 +1,240 @@
+change to the language.
+
+_Generics_: Generic tuple types raise similar issues as discussed
+for generic inline TypedDicts (above).
+
+## Lists, sets, and dicts representing their own types
+
+### Idea
+
+We could allow `[int]` instead of `list[int]`, `{int}` instead of
+`set[int]`, and `{str: int}` instead of `dict[str, int]`.
+
+### Problems
+
+_Conflict with TypedDict syntax_: The dict part of this idea conflicts
+with the suggested inline syntax for TypedDicts above. Not necessarily
+a problem since both are hypothetical, but it does mean that we can pick
+only one. (Why can't we allow both? `{"int": str}` could represent a
+`dict[int, str]`, because type expressions may be strings, or it could be
+a TypedDict with a single item.)
+
+_`|` operator_: This operator represents a union in the type system.
+However, sets and dicts already define this operator with conflicting
+semantics. Lists don't and we could in theory make `[int] | [str]` return
+`Union[[int], [str]]`, but adding that operation to such a basic type would
+be confusing for users.
+
+_Hashability_: Lists, sets, and dicts are not hashable. Various parts of the
+`typing.py` runtime implementation rely on hashability for efficiently
+deduplicating and caching typing-related objects. There are fallbacks for
+non-hashable types, but they are much slower.
+
+_Hashability, again_: The proposed syntax for set and dict literals requires
+the inner type to be hashable (only the key type for dicts). While most type
+forms are hashable, this is not universally true. For example, `Annotated`
+metadata may not be hashable.
+
+_Bad incentives_: This is a more subjective one. Style guidance often
+suggests using abstract, covariant types, such as `Sequence` or `Mapping`,
+instead of concrete, invariant ones like `list` or `dict`. Providing overly
+native-feeling syntax for list, set, and dict types goes against this advice.
+The most convenient option should be the best one.
+
+## Builtin functions as type expressions
+
+### Idea
+
+Three common typing-related objects have similar names to existing
+builtin functions. What if we allowed users to use the builtin functions
+in type expressions?
+
+```python
+x: any  # instead of typing.Any
+y: iter[str]  # instead of collections.abc.Iterable[str]
+z: callable[[str], int]  # instead of collections.abc.Callable[[str], int]
+```
+
+### Problems
+
+_Muddying the waters_: Functions are not valid in type expressions. Special-casing
+a few builtin functions for this would be confusing for users.
+
+_Implementation complexity_: For this idea to work, the `iter` and `callable`
+builtins would have to be made subscriptable. There is an open proposal
+to make all functions subscriptable ([PEP 718](https://peps.python.org/pep-0718/)),
+but failing that, we might have to add a new kind of builtin function just
+for these two. Not out of the question, but it makes the language core more
+complicated.
```

Comment:
Arguably this might actually conflict with PEP-718. PEP-718 would make it so that the meaning of a subscripted function would be to specialise the function. But under this idea, subscripting `callable` would not specialise the `callable` function (which is not generic)

Reply:
I did mean to say it would conflict, let me rewrite this for clarity.